[WIP] Cohort course-wide and inline discussion topics UI.

muzaffaryousaf opened the pull request:
This adds the UI to the cohort management tab on LMS side. It allows the staff to cohort the course-wide as well as, inline discussions via friendly UI. 

The Sandbox is available at: http://cohort-discussions.m.sandbox.edx.org/

symbolist commented on lms/djangoapps/django_comment_client/base/views.py:
```diff
@@ -27,7 +27,8 @@
     get_group_id_for_comments_service,
-    get_discussion_categories_ids
+    get_discussion_categories_ids,
+    get_discussion_category_map
```
Add a `,` at the end.

catong commented:
@muzaffaryousaf I think my previous sugggested wording for UI text in the Content-Specific Discussion Topics section was less than ideal. After chatting with @explorerleslie and @marcotuts about the intended behavior I think this revised wording will cause less confusion for users. My apologies for the change. @explorerleslie @marcotuts Do you think this new wording is helpful?

"Content-specific discussion topics are by default always divided by cohort. To make some content-specific discussion topics accessible to all learners regardless of cohort, clear Always cohort content-specific discussion topics, then clear the checkboxes for the individual discussion topics that you want to make unified.

Selecting Always cohort content-specific discussion topics overrides existing settings and makes all content-specific discussion topics divided by cohort."

marcotuts commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -1071,6 +1071,42 @@
+    }
+  }
+
+
```
For .wrapper-tabs &.subcategories, I would suggest using padding: 0 ($baseline*1.5); (this lines up the checkbox visually with the start of the text from the parent's li element text string.

explorerleslie commented:
@muzaffaryousaf we've just done a quick design huddle here. What we'd like to do is change the current checkbox to a radio button selector per some feedback from @marcotuts 

So, right now, we have 
[ X ] Always cohort content-specific discussion topics

We should change that to: 
( X ) Always cohort content-specific discussion topics
(  ) Cohort some content-specific discussion topics 

@catong should provide final copy for the above selections. 

The behavior should stay exactly the same otherwise, with disabling discussion topic checkboxes if the Always option is selected, etc. The only thing that changes is checkbox >> radio button. 

Let us know if you have questions.

catong commented:
@explorerleslie @muzaffaryousaf @marcotuts 
Replacement UI text to appear under the "Content-Specific Discussion Topics" heading and above the two new radio buttons:
"Specify whether content-specific topics are divided by cohort."
OK? Suggestions welcome.

explorerleslie commented:
looks good @catong

catong commented:
Oops; my mistake above: missing the word "discussion". The whole sentence should be:
"Specify whether content-specific discussion topics are divided by cohort."

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,48 @@
+                <div class="form-field">
+                    <div class="inline-discussion-topics">
+                        <h3 class="subsection-title"><%- gettext('Content-Specific Discussion Topics') %></h3>
+                        <p><%- gettext('Select the content-specific discussion topics that you want to divide by cohort. To specify that all existing and future content-specific discussion topics are divided by cohort, select ') %><b><%- gettext('Always cohort content-specific discussion topics') %></b>.</p>
```
Per conversations on this PR: replace this UI text with "Specify whether content-specific discussion topics are divided by cohort." 
along with new radio buttons

symbolist commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -362,3 +365,61 @@ def debug_cohort_mgmt(request, course_key_string):
+
+    discussions_category = get_discussion_category_map(course)
+    discussions_category['coursewide_categories'] = discussions_category['entries']
+    del discussions_category['entries']
```
You can do `discussions_category.pop('entries')` instead of the above two lines.

symbolist commented on lms/templates/instructor/instructor_dashboard_2/cohort_management.html:
```diff
@@ -10,6 +10,7 @@
      data-advanced-settings-url="${section_data['advanced_settings_url']}"
      data-upload_cohorts_csv_url="${section_data['upload_cohorts_csv_url']}"
      data-course_cohort_settings_url="${section_data['course_cohort_settings_url']}"
+     data-discussion_topics_url="${section_data['discussion_topics_url']}"
```
DOM attribute names generally have `-` and not `_`. Can you update this?

muzaffaryousaf commented:
@explorerleslie @catong @marcotuts i've addressed all of your comments and updated the code on sandbox, please have a look.
Thanks

muzaffaryousaf commented:
@symbolist can you please review this PR. Thanks.

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions.underscore:
```diff
@@ -0,0 +1,52 @@
+            <div class="form-fields">
+                <div class="form-field">
+                    <div class="course-wide-discussion-topics">
+                        <h3 class="subsection-title"><%- gettext('Course-wide Discussion Topics') %></h3>
```
@muzaffaryousaf My apologies for introducing this inconsistency!
"Course-wide" should have uppercase "W" as well: "Course-Wide Discussion Topics". Thanks!

catong commented:
I commented on one small UI label inconsistency. Other than that, it looks really good!

explorerleslie commented:
@marcotuts I feel like we have to hide the list of discussion topics when the "Always cohort" is selected. It's confusing to see the list of discussions, some of which are checked, some which aren't, some that have the "Cohorted" label, some that don't...when ALL of the topics are cohorted in this case. What do you think?

marcotuts commented:
@explorerleslie  - In this case I think we should leave the (now clearly disabled) discussions, since the disabled state makes clear that the above controls are associated with the options below. My primary concern with hiding the list would be that the discoverability of what the "Cohort some" option would let you do. I could be wrong here, but my suggestion might be to release this and just see whether people understand that the "Always / Some" setting is an override. Does that make sense?

catong commented:
@explorerleslie @marcotuts I'll make it as clear as I can in the documentation what course teams are seeing in the UI when they choose "Always cohort" after previously having only some discussion topics cohorted.

symbolist commented on lms/static/js/groups/views/cohort_discussions.js:
```diff
@@ -0,0 +1,208 @@
+            self.saveForm(self.$('.inline-discussion-topics'))
+                .done(function () {
+                    self.model.fetch().done(function () {
+                        self.render();
```
If I understand this correctly, doing this will re-render both the forms, right? And if so, the changes in the other form will be lost? Can we break the two forms into separate views so that only one is re-rendered?

"Working With Nested Models Or Collections" on http://addyosmani.github.io/backbone-fundamentals/ might be useful.

explorerleslie commented:
@marcotuts @catong ok, I'm fine with that approach. Let's see how it goes. Thanks for following up!

symbolist commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -89,7 +96,26 @@ def _get_cohort_representation(cohort, course):
 
 
-@require_http_methods(("GET", "PUT", "POST"))
+def get_cohorted_discussion_ids(course, cohorted_discussions):
```
I think it would make more sense to call this function `get_cohorted_discussions()` and pass it `cohort_settings` so that it can pick the `cohorted_discussions` from it itself.

symbolist commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -99,15 +125,40 @@ def course_cohort_settings_handler(request, course_key_string):
+        )
+
+        if request.json.get('cohorted_course_wide_discussions'):
+            cohorted_inline_ids.extend(request.json.get('cohorted_course_wide_discussions'))
```
Should we also not validate that the ids are valid?

symbolist commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -99,15 +125,40 @@ def course_cohort_settings_handler(request, course_key_string):
+            return JsonResponse({"error": unicode(err)}, 400)
+    elif request.method in ["PUT", "POST"]:
         is_cohorted = request.json.get('is_cohorted')
         if is_cohorted is None:
```
Using `PUT` or `POST` just for setting `is_cohorted` is no longer consistent with the semantics of these methods. This should be changed to use `PATCH` too.

symbolist commented on lms/templates/instructor/instructor_dashboard_2/cohort-discussions-inline.underscore:
```diff
@@ -0,0 +1,30 @@
+                    <p><%- gettext('Specify whether content-specific discussion topics are divided by cohort.') %></p>
+                    <div class="always_cohort_inline_discussions">
+                        <input type="radio" name="inline" class="check-all-inline-discussions" <%- alwaysCohortInlineDiscussions ? 'checked="checked"' : '' %>/>
+                        <span class="all-inline-discussions"><%= interpolate('Always cohort content-specific discussion topics') %></span>
```
@catong Shouldn't "Always cohort content-specific discussion topics" be "Cohort all content-specific discussion topics" to mirror "Cohort some content-specific discussion topics"?

catong replied:
@symbolist Originally, we wanted to reflect exactly the same name as the flag was named in Advanced Settings so that users would understand that the behavior of that control had not changed. The word "Always" was meaningful in that it highlighted that future as well as existing inline discussion topics were affected by this setting. It's true that "Cohort all... topics" would be more parallel to "Cohort some...topics". Perhaps that wording does adequately capture that all content-specific discussion topics created after selecting this option are also cohorted. @explorerleslie @marcotuts what are your thoughts? If you all think that we don't lose nuances by moving away from "Always cohort..." then I agree with the change.

explorerleslie replied:
@symbolist while I understand your desire for parallelism here, I agree with @catong that "Always cohort" is stronger and also implies that any new discussion topics created will also be cohorted. I would like to keep it as is.

symbolist commented on lms/static/js/groups/views/cohort_discussions_course_wide.js:
```diff
@@ -0,0 +1,84 @@
+         Returns:
+            HTML list for course-wide discussion topics.
+        **/
+        getCourseWideDiscussions: function (courseWideDiscussions) {
```
`getCourseWideDiscussionsHtml()` would be more accurate.

symbolist commented on lms/static/js/groups/views/cohort_discussions_course_wide.js:
```diff
@@ -0,0 +1,84 @@
+        /**
+         Enables the save button for course-wide discussions.
+        **/
+        changeCourseWideDiscussionCategory: function(event) {
```
How about `discussionCategoryStateChanged`?

symbolist commented on lms/static/js/groups/views/cohorts.js:
```diff
@@ -277,6 +279,27 @@ var edx = edx || {};
+
+            if (!this.CourseWideDiscussionsView) {
+                this.CourseWideDiscussionsView = new CourseWideDiscussionsView({
+                    el: cohortDiscussionsElement,
```
Is there any reason we need to pass the parent element instead of the elements for the sections?

symbolist commented on lms/static/js/groups/views/cohort_discussions_inline.js:
```diff
@@ -0,0 +1,142 @@
+                alwaysCohortInlineDiscussions:alwaysCohortInlineDiscussions
+            }));
+
+            $('ul.inline-topics').qubit();
```
Can you document the purpose of this?

symbolist commented on lms/static/js/groups/views/cohort_discussions_inline.js:
```diff
@@ -0,0 +1,142 @@
+         Returns:
+            HTML list for inline discussion topics.
+        **/
+        getInlineDiscussions: function (inlineDiscussions) {
```
This can also be `getInlineDiscussionsHtml()`.

symbolist commented on lms/static/js/groups/views/cohort_discussions_inline.js:
```diff
@@ -0,0 +1,142 @@
+         Enables the save button for inline discussion topics.
+
+         Args:
+            disable (Bool): The flag to enable/disable the elements.
```
It does not look like this is applicable here.

symbolist commented on lms/static/js/groups/views/cohort_discussions_inline.js:
```diff
@@ -0,0 +1,142 @@
+         Args:
+            disable (Bool): The flag to enable/disable the elements.
+        **/
+        changeAllInlineDiscussions: function(event) {
```
I think `setCohortAllInlineDiscussions` and `setCohortSomeInlineDiscussions` would be more descriptive names.

symbolist commented on lms/static/js/groups/views/cohort_discussions.js:
```diff
@@ -0,0 +1,113 @@
+         Returns the cohorted discussions list.
+
+         Args:
+            selector (HTML element): The topic element to get the ids.
```
How about "selector (string): To select the discussion elements whose ids to return."?

symbolist commented on lms/static/js/groups/views/cohort_discussions_inline.js:
```diff
@@ -0,0 +1,142 @@
+         Args:
+            disable (Bool): The flag to enable/disable the elements.
+        **/
+        toggleInlineDiscussions: function(disable) {
```
You can make this `setElementsEnabled: function(enable_checkboxes, enable_save_button)` and then use it in `render()` too.

symbolist commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -362,3 +406,41 @@ def debug_cohort_mgmt(request, course_key_string):
+    The handler for cohort discussion categories requests.
+    This will raise 404 if user is not staff.
+
+    Returns the JSON representation of discussion topics w.r.t categories for the course.
```
Can you add an example of the return value? For example, see: https://github.com/edx/edx-ora2/blob/master/openassessment/xblock/data_conversion.py#L10-L63

catong commented:
@muzaffaryousaf Has the sandbox http://cohort-discussions.m.sandbox.edx.org/ expired? Can you please refresh it if so? Thanks!

muzaffaryousaf commented:
@catong i've updated the sandbox. Thanks

catong commented:
@muzaffaryousaf Thanks!

muzaffaryousaf commented:
test this please

muzaffaryousaf commented:
@symbolist @explorerleslie @catong i've addressed all the comments on this PR. I'm going to re-open this PR against **ammar/move-cohort-management-tnl-1268**  as we have all cohort related functionality in that brach.
Thanks.

muzaffaryousaf commented:
The PR would be this https://github.com/edx/edx-platform/pull/7321.
